language/borrow-graph/src/graph.rs

```diff
@@ -0,0 +1,349 @@
+    // Edges/Borrows
+    //**********************************************************************************************
+
+    /// Add a strong (exact) epsilon borrow from `parent_id` to `child_id`
+    pub fn add_strong_borrow(&mut self, loc: Loc, parent_id: RefID, child_id: RefID) {
+        self.factor(parent_id, loc, vec![], child_id)
+    }
+
+    /// Add a strong (exact) field borrow from `parent_id` to `child_id` at field `field`
+    pub fn add_strong_field_borrow(
+        &mut self,
+        loc: Loc,
+        parent_id: RefID,
+        field: Lbl,
+        child_id: RefID,
+    ) {
+        self.factor(parent_id, loc, vec![field], child_id)
+    }
+
+    /// Add a weak (prefix) epsilon borrow from `parent_id` to `child_id`
+    /// i.e. `child_id` might be borrowing from ANY field in `parent_id`
+    pub fn add_weak_borrow(&mut self, loc: Loc, parent_id: RefID, child_id: RefID) {
+        self.add_path(parent_id, loc, false, vec![], child_id)
+    }
+
+    /// Add a weak (prefix) field borrow from `parent_id` to `child_id` at field `field`
+    /// i.e. `child_id` might be borrowing from ANY field in `parent_id` rooted at `field`
+    pub fn add_weak_field_borrow(
+        &mut self,
+        loc: Loc,
+        parent_id: RefID,
+        field: Lbl,
+        child_id: RefID,
+    ) {
+        self.add_path(parent_id, loc, false, vec![field], child_id)
+    }
+
+    fn add_edge(&mut self, parent_id: RefID, edge: BorrowEdge<Loc, Lbl>, child_id: RefID) {
+        let parent = self.0.get_mut(&parent_id).unwrap();
+        parent
+            .borrowed_by
+            .0
+            .entry(child_id)
+            .or_insert_with(BTreeSet::new)
+            .insert(edge);
+        let child = self.0.get_mut(&child_id).unwrap();
+        child.borrows_from.insert(parent_id);
+    }
+
+    fn add_path(
+        &mut self,
+        parent_id: RefID,
+        loc: Loc,
+        strong: bool,
+        path: Path<Lbl>,
+        child_id: RefID,
+    ) {
+        let edge = BorrowEdge { strong, path, loc };
+        self.add_edge(parent_id, edge, child_id)
+    }
+
+    fn factor(&mut self, parent_id: RefID, loc: Loc, path: Path<Lbl>, intermediate_id: RefID) {
+        let parent = self.0.get_mut(&parent_id).unwrap();
+        let mut needs_factored = vec![];
+        for (child_id, parent_to_child_edges) in &parent.borrowed_by.0 {
+            for parent_to_child_edge in parent_to_child_edges {
+                if paths::leq(&path, &parent_to_child_edge.path) {
+                    let factored_edge = (*child_id, parent_to_child_edge.clone());
+                    needs_factored.push(factored_edge);
+                }
+            }
+        }
+
+        let mut cleanup_ids = BTreeSet::new();
+        for (child_id, parent_to_child_edge) in &needs_factored {
+            let parent_to_child_edges = parent.borrowed_by.0.get_mut(child_id).unwrap();
+            assert!(parent_to_child_edges.remove(parent_to_child_edge));
```

Comment:
What does having executable code inside an assert! mean?  Could it get compiled away under some circumstances?

Reply:
We should write down some invariants describing the consistency of the borrow graph and check them at entry and exit of every public API that mutates the borrow graph, especially ones that are as complicated as this factor.

Reply:
> What does having executable code inside an assert! mean? Could it get compiled away under some circumstances?

It will always be compiled. `debug_assert` is used for conditional checks